pkg/image/index/index.go

```diff
@@ -0,0 +1,179 @@
+package index
+
+import (
+	"time"
+
+	"github.com/docker/distribution/digest"
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/client"
+	imageapi "github.com/openshift/origin/pkg/image/api"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type ImageIndex struct {
+	client client.ImageInterface
+
+	imageController  *framework.Controller
+	imageIndex       cache.Indexer
+	imageStoreSynced func() bool
+}
+
+var (
+	ChainIndexName     = "chain"
+	LastChainIndexName = "last-chain"
+	// TODO: Re-listing images might be expensive on large clusters, this should be
+	// configurable?
+	ResyncInterval = 10 * time.Minute
+)
+
+// NewImageIndex returns a new image index.
+func NewImageIndex(client client.ImageInterface, stopChan <-chan struct{}) *ImageIndex {
+	c := &ImageIndex{
+		client: client,
+	}
+	c.imageIndex, c.imageController = framework.NewIndexerInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+				return c.client.List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return c.client.Watch(options)
+			},
+		},
+		&imageapi.Image{},
+		ResyncInterval,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    c.addImage,
+			UpdateFunc: c.updateImage,
+			DeleteFunc: c.deleteImage,
+		},
+		cache.Indexers{
+			ChainIndexName:     chainIndexFunc,
+			LastChainIndexName: lastChainIndexFunc,
+		},
+	)
+
+	go c.imageController.Run(stopChan)
+	return c
+}
+
+func (c *ImageIndex) WaitForSyncedStores() {
+	for !c.imageController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+// Descendants provides a list of images that the provided image is based on.
+// Note that if a base image is tagged multiple times, this will return all the matching
+// images.
+func (c *ImageIndex) Descendants(image *imageapi.Image) ([]*imageapi.Image, error) {
```

Comment:
@miminar it depends on how you look at the image tree... the "child" images might seems more accurate but they are not really, because you can't say that these are direct children (as there might be more "parents" which represents the same image but with different tag). In the tree, the ancestors are all images that are above the image and descendads are the images below that image (or in other words, descendants are the the "base" images and the "ancestors" are images that are based on this "base" image).